src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -568,6 +568,54 @@ public ML_DSA_KeyPair generateKeyPairInternal(byte[] randomBytes) {
         return new ML_DSA_KeyPair(sk, pk);
     }
 
+    private static int[][] deepClone(int[][] array) {
+        int[][] clone = new int[array.length][];
+        for (int i = 0; i < array.length; i++) {
+            clone[i] = array[i].clone();
+        }
+        return clone;
+    }
+
+    // This is similar to the generateKeyPairInternal method. Instead of
+    // generating from a seed, it uses stored fields inside the private key
+    // to calculate the public key. It performs several checks during the
+    // calculation to make sure the private key is a valid one. Otherwise,
+    // an IllegalArgumentException is thrown.
+    public ML_DSA_PublicKey privKeyToPubKey(ML_DSA_PrivateKey sk) {
+        // Sample A
+        int[][][] keygenA = generateA(sk.rho); //A is in NTT domain
+
+        // Compute t and tr
+        // make a copy of sk.s1 and modify it. Although we can also
+        // take it out of NTT domain later, it was modified for a while.
+        var s1 = deepClone(sk.s1);
+        mlDsaVectorNtt(s1); //s1 now in NTT domain
+        int[][] As1 = new int[mlDsa_k][ML_DSA_N];
```

Comment:
Multiarray allocations like this are quite slow - for better performance consider using the `integerMatrixAlloc` method. This also applies to the allocations on L598 and L599